state/validation.go

```diff
@@ -132,7 +132,7 @@ func validateBlock(evidencePool EvidencePool, stateDB dbm.DB, state State, block
 
 	// Validate all evidence.
 	for _, ev := range block.Evidence.Evidence {
-		if err := VerifyEvidence(stateDB, state, ev); err != nil {
+		if err := VerifyEvidence(stateDB, state, ev, &block.Header); err != nil {
```

Comment:
If the evidence is of the type Lunatic Validator then we will need to retrieve the header from the nodes block store in order to validate it else this check will fail everytime.

Reply:
As a side note, I think here we should also first check that have the evidence before trying to verify it again. That way we save time because we just need to check that it hasn't expired (or that the validator is still punishable). I can create an issue for it.

Reply:
> If the evidence is of the type Lunatic Validator then we will need to retrieve the header from the nodes block store in order to validate it else this check will fail everytime.

we have the block here (we're validating it, so the header is always present).

> As a side note, I think here we should also first check that we have the evidence already stored before trying to verify it again

You're assuming that the node has all evidence, which might not be the case! But I agree, we should probably make this optimization! I'll create an issue.

Reply:
https://github.com/tendermint/tendermint/issues/4696

Reply:
> You're assuming that the node has all evidence, which might not be the case! But I agree, we should probably make this optimization! I'll create an issue.

Yeah I just meant it as an optimization thing